app/experimenter/experiments/tests/test_views.py

```diff
@@ -298,6 +299,37 @@ def test_list_view_filters_and_orders_experiments(self):
             list(context["experiments"]), list(filtered_ordered_experiments)
         )
 
+    def test_list_view_total_experiments_count(self):
+        user_email = "user@example.com"
+
+        number_of_experiments = settings.EXPERIMENTS_PAGINATE_BY + 1
+        for i in range(number_of_experiments):
+            ExperimentFactory.create_with_status(
+                random.choice(Experiment.STATUS_CHOICES)[0]
+            )
+
+        response = self.client.get(
+            reverse("home"), **{settings.OPENIDC_EMAIL_HEADER: user_email}
+        )
+        self.assertEqual(response.status_code, 200)
+        html = response.content.decode("utf-8")
```

Comment:
From skimming the rest of the file, all other tests, that use `self.client` don't test the final output but instead test the content (aka. `context`) of the variables sent into the HTML rendering. My test is, as far as I can see, the first to actually study the generated HTML. The reason for this strong. Had I just tested the `response.context` I would not have been able to test that the Django template rendering did the right thing. All I'd find is the `paginator` object. 

Also, in this test I convert the response content to a string and use regular expressions. It works but regular expressions never feel great since they can be fragile. 

I think my test, as it is now, is very much a scripted way of what a human would do by rendering the HTML in a browser and looking at the output.

Reply:
I wouldn't mind using PyQuery! That's what we did on crashstats in the socorro project to investigate what rendered in the Jinja templates. 

Mind you, had this been PyQuery it would still have to look something like this:

```python
doc = PyQuery(response.content.decode("utf-8"))
for header in doc('h2.title').items():
    assert re.findall(r'\d+\s+Experiments', header.text())
```
Since the string "21 Experiments" isn't in the generated HTML.

Reply:
@peterbe @mythmon In all the time that I've been a Django developer I have generally steered clear of validating HTML in unit tests becuase it can change so frequently and I don't think that Django supports very well developed patterns of cleanly testing it.  In React it's much more idiomatic to test and validate the output of individual components, that's one of the reasons I like it, and maybe someday we'll migrate the frontend of Experimenter to React and can test the UI more rigourously.  In the meantime, I think it's fine that you've added this test here but I don't think it's worth expanding this practice to the project in general, I think just checking in the browser, including screenshots in PRs, and letting QA pass over things is sufficient.

Reply:
A couple of points...

I think I made a mistake when using `response.content.decode("utf-8")` to get to the HTML as a Unicode string. The proper way is `content = response.content; content = content.decode(response.charset)`. Because that's how `SimpleTestCase._assert_contains` does it. 

The QA that my test added would not have been possible because the critical bug was in the *template*. Not in the *view*. E.g.

```python
def myview(request):
    return render(request, 'template.html', foo="bar")
```
```html
<!-- template.html -->
<body>{{ fooo }}</body>
```
Had you done `self.assertEqual(response.context['foo'], 'bar')` you might have missed the typo in the the template. 

Yes, testing the output is fragile but testing implementation details are even worse :)